File: src/core/cache.js
/**
 * MTSCOS AI 系统 - 缓存模块
 * 用于缓存系统数据，提高性能
 */

class Cache {
    constructor() {
        this.cache = new Map();
        this.defaultTTL = 3600; // 默认过期时间1小时
    }
    
    // 设置缓存项
    set(key, value, ttl = this.defaultTTL) {
        const expiresAt = Date.now() + (ttl * 1000);
        this.cache.set(key, { value, expiresAt });
    }
    
    // 获取缓存项
    get(key) {
        const item = this.cache.get(key);
        if (!item) {
            return null;
        }
        
        // 检查是否过期
        if (Date.now() > item.expiresAt) {
            this.cache.delete(key);
            return null;
        }
        
        return item.value;
    }
    
    // 删除缓存项
    delete(key) {
        return this.cache.delete(key);
    }
    
    // 清空所有缓存
    clear() {
        this.cache.clear();
    }
    
    // 检查缓存项是否存在
    has(key) {
        return this.get(key) !== null;
    }
    
    // 获取缓存大小
    size() {
        // 清理过期项
        this.cleanup();
        return this.cache.size;
    }
    
    // 清理过期缓存项
    cleanup() {
        const now = Date.now();
        for (const [key, item] of this.cache.entries()) {
            if (now > item.expiresAt) {
                this.cache.delete(key);
            }
        }
    }
}

module.exports = new Cache();

File: flask-app/app/utils/cache.py
# -*- coding: utf-8 -*-
#!/usr/bin/env python3
"""缓存工具模块"""

import time
import functools
import logging
from typing import Any, Dict, Optional, Callable

logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger('cache')

class CacheManager:
    """缓存管理器"""

    def __init__(self, default_ttl: int = 3600):
        """初始化缓存管理器

        Args:
            default_ttl: 默认缓存过期时间(秒)
        """
        self.cache: Dict[str, Dict[str, Any]] = {}
        self.default_ttl = default_ttl
        self.hits = 0
        self.misses = 0

    def _generate_key(self, func: Callable, args: tuple, kwargs: dict) -> str:
        """生成缓存键

        Args:
            func: 函数对象
            args: 函数参数
            kwargs: 函数关键字参数

        Returns:
            缓存键
        """
        key_parts = [
            func.__module__,
            func.__name__,
            str(args),
            str(sorted(kwargs.items()))
        ]
        return "::".join(key_parts)

    def get(self, key: str) -> Optional[Any]:
        """获取缓存值

        Args:
            key: 缓存键

        Returns:
            缓存值,如果不存在或过期返回None
        """
        if key not in self.cache:
            return None

        item = self.cache[key]
        if time.time() > item['expiry']:
            self.misses += 1
            return None

        self.hits += 1
        return item['value']

    def set(self, key: str, value: Any, ttl: Optional[int] = None) -> None:
        """设置缓存值

        Args:
            key: 缓存键
            value: 缓存值
            ttl: 过期时间(秒)
        """
        ttl = ttl or self.default_ttl
        self.cache[key] = {
            'value': value,
            'expiry': time.time() + ttl
        }

    def delete(self, key: str) -> None:
        """删除缓存

        Args:
            key: 缓存键
        """
        if key in self.cache:
            del self.cache[key]

    def clear(self) -> None:
        """清空缓存"""
        self.cache.clear()
        self.hits = 0
        self.misses = 0

    def size(self) -> int:
        """获取缓存大小

        Returns:
            缓存项数量
        """
        return len(self.cache)

    def stats(self) -> Dict[str, Any]:
        """获取缓存统计信息

        Returns:
            统计信息
        """
        return {
            'hits': self.hits,
            'misses': self.misses,
            'total': self.hits + self.misses,
            'size': self.size()
        }

    def decorator(self, ttl: Optional[int] = None):
        """缓存装饰器

        Args:
            ttl: 过期时间(秒)

        Returns:
            装饰器函数
        """
        def decorator(func: Callable) -> Callable:
            @functools.wraps(func)
            def wrapper(*args, **kwargs):
                key = self._generate_key(func, args, kwargs)
                cached_value = self.get(key)

                if cached_value is not None:
                    logger.debug(f"缓存命中: {key}")
                    return cached_value

                logger.debug(f"缓存未命中: {key}")
                result = func(*args, **kwargs)
                self.set(key, result, ttl)
                return result

            return wrapper

        return decorator

cache_manager = CacheManager()

def get_cache_manager():
    """获取缓存管理器实例"""
    return cache_manager